File: src/main/java/com/imyyq/mvvm/binding/viewadapter/listview/ViewAdapter.java
package com.imyyq.mvvm.binding.viewadapter.listview;

import android.widget.AbsListView;
import android.widget.ListView;

import androidx.databinding.BindingAdapter;

import com.imyyq.mvvm.binding.command.BindingCommand;

public final class ViewAdapter {

    @BindingAdapter(value = {"onScrollChangeCommand", "onScrollStateChangedCommand"}, requireAll = false)
    public static void onScrollChangeCommand(final ListView listView,
                                             final BindingCommand<ListViewScrollDataWrapper> onScrollChangeCommand,
                                             final BindingCommand<Integer> onScrollStateChangedCommand) {
        listView.setOnScrollListener(new AbsListView.OnScrollListener() {
            private int scrollState;

            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
                this.scrollState = scrollState;
                if (onScrollStateChangedCommand != null) {
                    onScrollStateChangedCommand.execute(scrollState);
                }
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
                if (onScrollChangeCommand != null) {
                    onScrollChangeCommand.execute(new ListViewScrollDataWrapper(scrollState, firstVisibleItem, visibleItemCount, totalItemCount));
                }
            }
        });

    }


    @BindingAdapter(value = {"onItemClickCommand"}, requireAll = false)
    public static void onItemClickCommand(final ListView listView, final BindingCommand<Integer> onItemClickCommand) {
        listView.setOnItemClickListener((parent, view, position, id) -> {
            if (onItemClickCommand != null) {
                onItemClickCommand.execute(position);
            }
        });
    }

    public static class ListViewScrollDataWrapper {
        public final int firstVisibleItem;
        public final int visibleItemCount;
        public final int totalItemCount;
        public final int scrollState;

        public ListViewScrollDataWrapper(int scrollState, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
            this.firstVisibleItem = firstVisibleItem;
            this.visibleItemCount = visibleItemCount;
            this.totalItemCount = totalItemCount;
            this.scrollState = scrollState;
        }
    }
}

File: src/main/java/com/imyyq/mvvm/binding/viewadapter/listview/ViewAdapter.kt
package com.imyyq.mvvm.binding.viewadapter.listview

import android.widget.AbsListView
import android.widget.ListView
import androidx.databinding.BindingAdapter
import com.imyyq.mvvm.binding.command.BindingConsumer

@BindingAdapter(
    value = ["onScrollChangeCommand", "onScrollStateChangedCommand"],
    requireAll = false
)
fun onScrollChangeCommand(
    listView: ListView,
    onScrollChangeCommand: BindingConsumer<ListViewScrollDataWrapper>?,
    onScrollStateChangedCommand: BindingConsumer<Int>?
) {
    listView.setOnScrollListener(object : AbsListView.OnScrollListener {
        private var scrollState = 0
        private lateinit var wrapper: ListViewScrollDataWrapper
        override fun onScrollStateChanged(
            view: AbsListView,
            scrollState: Int
        ) {
            this.scrollState = scrollState
            onScrollStateChangedCommand?.call(scrollState)
        }

        override fun onScroll(
            view: AbsListView,
            firstVisibleItem: Int,
            visibleItemCount: Int,
            totalItemCount: Int
        ) {
            if (onScrollChangeCommand != null) {
                if (!this::wrapper.isInitialized) {
                    wrapper = ListViewScrollDataWrapper()
                }
                wrapper.firstVisibleItem = firstVisibleItem
                wrapper.visibleItemCount = visibleItemCount
                wrapper.totalItemCount = totalItemCount
                wrapper.scrollState = scrollState
                onScrollChangeCommand.call(wrapper)
            }
        }
    })
}

class ListViewScrollDataWrapper {
    var firstVisibleItem = 0
    var visibleItemCount = 0
    var totalItemCount = 0
    var scrollState = 0
    override fun toString(): String {
        return "ListViewScrollDataWrapper{" +
                "firstVisibleItem=" + firstVisibleItem +
                ", visibleItemCount=" + visibleItemCount +
                ", totalItemCount=" + totalItemCount +
                ", scrollState=" + scrollState +
                '}'
    }
}